src/ucs/arch/atomic.h

```diff
@@ -16,6 +16,7 @@
 #elif defined(__aarch64__)
 #  include "generic/atomic.h"
 #else
+/* cppcheck-suppress preprocessorErrorDirective */
```

Comment:
Gross.

Reply:
BTW, Azure can run aarch64 images (through binfmt_misc)

Reply:
That is what we do for all `cppcheck` supressions. This one is rather specific, we can define one of the variants (`__aarch64__`, `__x86_64__` or `__powerpc64__`) but that reduces coverage.

Reply:
cppcheck is really limited, and adding gorss comments all over the place to shut it up is a poor practice.

Reply:
We need its coverage and our current approach is using inline suppressions. Could you please suggest another way of doing that without loosing coverage from `cppcheck` (whatever quality it is)?

Reply:
I would not expect cppcheck to be clean in CI, its has too poor quality.

cppcheck is a tool like coverity that you run occasionally, wade through its endless false positives and try to find some value.

Once you start annotating source like this you'll have hundreds of cases and a messy source tree. Next people start preemptively/wrongly adding the annotations and the utility of the static tools is wrecked.

This is a great example, because cppcheck should NOT be reporting the above error, it means that it was invoked improperly with the wrong set of preprocessor symbols defined. ie this is not a false positive at all, and really indicates a CI bug.

Reply:
RHEL team runs `cppcheck` and CI should help us identify issues before RH verification finds it.

Reply:
BTW could we run cppcheck with __x86_64__?
see this : https://www.systutorials.com/docs/linux/man/1-cscppc/

Reply:
@yosefe yes, we can do that, but that reduces coverage. Alternatively, we can run it three times, setting all expected defines. Pure `cppcheck` (not `cscppc`) permutes macros' values to check as many code paths as it can.

Reply:
we should probably take arch from the system (e.g uname -m) or somehow extract it from gcc
can we even run cross-arch static check .. ?

Reply:
I don't see any way to set "at least one of" arch macros defined on one cmd, `cppcheck` permutes all options. We can explicitly set/unset any of them which leads to lesser coverage.

RH's `cscppc` checks only `__x86_64__` according to its source:
https://github.com/kdudka/cscppc/blob/1b1c0e1fd0e83b20ccb7f968113de50e79cc85ac/cscppc.c#L45-L55.

Reply:
can we run 3 commands with -D for each of the archs we support?

Reply:
> RHEL team runs `cppcheck` and CI should help us identify issues before RH verification finds it.

I cannot stress this enough, you have the wrong impression of RHEL. They run cppcheck and coverity and look for obvious hits. They do not expect any package to be cppcheck and coverity clean.

It is reasonable to run only once with __x86_64__ unless ucx has a lot of arch specific code.